### PR TITLE
New feature: blendColors for bubble series

### PR DESCRIPTION
**Pull request by madepiet**
Added new feature for a `bubble` series, [blendColors](https://api.highcharts.com/highcharts/plotOptions.bubble.blendColors). See [the bubble series docs article](https://www.highcharts.com/docs/chart-and-series-types/bubble-series#blend-colors) for more info.

### CODE REVIEW

File: ts/Series/TemperatureMap/TemperatureMapSeries.ts [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+                            (
+                                (options.marker || {}).enabled &&
+                                typeof pointMarkerOptions.enabled ===
+                                    'undefined'
+                            ) || pointMarkerOptions.enabled
+                        ) &&
+                        !point.isNull &&
+                        point.visible !== false &&
+                        // Above zThreshold
+                        (point.options as any).z >= (this.options.zMin || 0);
```
Comment (raf18seb) [inline ts/Series/TemperatureMap/TemperatureMapSeries.ts:291]: The only one `as any` I couldn't get rid of. Any suggestions?

File: ts/Series/MapBubble/MapBubbleSeries.ts [hunk 2]
```diff
@@ -273,7 +275,9 @@ class MapBubbleSeries extends BubbleSeries {
     translate(): void {
         MapPointSeries.prototype.translate.call(this);
         this.getRadii();
+        (this as any).adjustMinSize = true;
         this.translateBubble();
+        (this as any).adjustMinSize = false;
```
Comment (raf18seb) [inline ts/Series/MapBubble/MapBubbleSeries.ts:280]: Here (inside `translateBubble`) I need to fire modified (modified by this `adjustMinSize` flag) `getPxExtremes` method.
Any ideas on how can I do it better?

**Review by raf18seb — COMMENTED** [review]
I have 2 questions, see the comments below

File: ts/Series/TemperatureMap/TemperatureMapSeries.ts (hunk 1 above)
Comment (TorsteinHonsi) [reply ts/Series/TemperatureMap/TemperatureMapSeries.ts:291]: `point.options.z` is optional, so you need to test if it is a number too. Also make sure it works with array point configs. Maybe you should check `point.z` instead?

Comment (raf18seb) [reply ts/Series/TemperatureMap/TemperatureMapSeries.ts]: Oops, I saw a different error, not sure what happened. Of course it's optional. Fixed now, thanks!
ps. works fine with array point and `series.keys` config

**Review by DJTechnoo — APPROVED** [review]

**Review by TorsteinHonsi — COMMENTED** [review]
Is lon/lat not supported yet or am I doing something wrong? https://jsfiddle.net/highcharts/h9gqszo7/

**Comment by KacperMadej** [comment]
@TorsteinHonsi 
The coordinates are outside of Europe. Works fine on a world map: https://jsfiddle.net/BlackLabel/0gp4ua6r/

File: ts/Series/TemperatureMap/TemperatureMapPoint.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+/* *
+ *
+ *  Imports
+ *
+ * */
+
+import type TemperatureMapSeries from './TemperatureMapSeries.js';
+import type TemperatureMapPointOptions from './TemperatureMapPointOptions';
+
+import MapBubblePoint from '../MapBubble/MapBubblePoint.js';
```
Comment (KacperMadej) [inline ts/Series/TemperatureMap/TemperatureMapPoint.ts:22]: Pulling series and points directly is forcing the builder to include those parts in the module. Use `SeriesRegistry` as an alternative.

File: ts/Series/TemperatureMap/TemperatureMapSeries.ts [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+     *  Functions
+     *
+     */
+
+    /**
+     * @private
+     */
+    public isPointInside(): true {
+        // Do not hide cut bubbles outside the plot area
+        return true;
```
Comment (KacperMadej) [inline ts/Series/TemperatureMap/TemperatureMapSeries.ts:123]: Is there any better way? Should this be a more general (bubble series level problem) problem for later?

File: ts/Series/TemperatureMap/TemperatureMapSeries.ts [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+    public getPxExtremes(): {
+        minPxSize: number,
+        maxPxSize: number
+    } {
+        const ret = super.getPxExtremes.apply(this),
+            temperatureColors = this.options.temperatureColors,
+            sizeFactor = this.sizeFactor || (1 / temperatureColors.length);
+
+        // Recalculate minPxSize to draw smaller bubbles
+        if (this.adjustMinSize) { // only when necessary
```
Comment (KacperMadej) [inline ts/Series/TemperatureMap/TemperatureMapSeries.ts:138]: When is it necessary? Could you add more info?

File: ts/Series/TemperatureMap/TemperatureMapSeries.ts [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+    }
+
+    /**
+     * Perform animation on the bubbles
+     * @private
+     */
+    public animate(init?: boolean): void {
+        if (
+            !init &&
+            this.points.length < (this.options.animationLimit as any) // #8099
```
Comment (KacperMadej) [inline ts/Series/TemperatureMap/TemperatureMapSeries.ts:186]: Will `pick(this.options.animationLimit, Infinity)` do the same without `any` casting?

File: ts/masters/modules/temperaturemap.src.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
+/**
+ * @license Highcharts JS v@product.version@ (@product.date@)
+ * @module highcharts/modules/temperaturemap
+ * @requires highcharts
+ *
+ * (c) 2009-2021
```
Comment (KacperMadej) [inline ts/masters/modules/temperaturemap.src.ts:6]: 2022

File: ts/Series/TemperatureMap/TemperatureMapSeries.ts [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+             * the smallest top marker is 5 times smaller than the biggest
+             * bottom one).
+             *
+             * @sample maps/plotoptions/temperaturemap-themes
+             *         Temperature colors themes
+             * @sample maps/plotoptions/temperaturemap-color-steps
+             *         Temperature colors with steps
+             *
+             * @default ["#ff0000", "#ffff00", "#00ff00", "#00ffff", "#0000ff"]
+             * @type {Array<ColorType>}
```
Comment (KacperMadej) [inline ts/Series/TemperatureMap/TemperatureMapSeries.ts:100]: The second type (probably `Array<[number,ColorType]>`) is not included in the API docs.

**Comment by TorsteinHonsi** [comment]
Status: The team decided that this should rather be implemented as a feature of the Bubble (and by extension, the MapBubble) series. Should be attempted implemented as a core feature in BubbleSeries, alternatively as a module should we decide that it bloats `highcharts-more.js`.

**Comment by raf18seb** [comment]
I'm already working on that ☝️ 
Converting to Draft PR until refactoring is done.

File: docs/maps/temperaturemap-series.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
+Temperature map
```
Comment (TorsteinHonsi) [inline docs/maps/temperaturemap-series.md:1]: Should this file be deleted now that we moved the feature over to standard bubble?

File: docs/sidebars.js [hunk 10]
```diff
@@ -62,6 +62,7 @@ module.exports = {
             'maps/mapline-series',
             'maps/mappoint-series',
             'maps/mapbubble-series',
+            'maps/temperaturemap-series',
```
Comment (TorsteinHonsi) [inline docs/sidebars.js:65]: As above

File: samples/maps/demo/blend-colors/readme.md [hunk 11]
```diff
@@ -0,0 +1,2 @@
+#  Temperature map
```
Comment (TorsteinHonsi) [inline samples/maps/demo/blend-colors/readme.md:1]: This description is obsolete, right?

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Thanks, looking good. Some change requests dealing with the demos are added.

Also, please update the main description at the top as well - the one that goes into the changelog.

File: docs/chart-and-series-types/bubble-series.md [hunk 12]
```diff
@@ -61,6 +61,45 @@ In the following example, a color axis is added and the `x` value is used for co
 
 <iframe style="width: 100%; height: 480px; border: none;" src="https://www.highcharts.com/samples/embed/highcharts/coloraxis/changed-default-color-key" allow="fullscreen"></iframe>
 
+Blend colors
+----------
+It is also possible to show another dimention - density. By using the [bubble.blendColors](https://api.highcharts.com/highcharts/series.bubble.blendColors) option, you can add a bubble with multiple graphics (colors) where each color graphic has the same zIndex as all other graphics with the same colors. This allows to show a nice blend effect that could be very useful for showing density ([espetially in a map chart](https://www.highcharts.com/docs/maps/mapbubble-series#mapbubble-blend-colors)).
```
Comment (DJTechnoo) [inline docs/chart-and-series-types/bubble-series.md:66]: A little spelling error: dimention -> dimension

Comment (TorsteinHonsi) [reply docs/chart-and-series-types/bubble-series.md:66]: I just fixed this and a few other minor things, will push soon.

File: ts/Series/Bubble/BubbleSeries.ts [hunk 13]
```diff
@@ -528,32 +552,45 @@ class BubbleSeries extends ScatterSeries {
...
-                        this.options.animation
-                    );
-                }
-            }, this);
+        if (!init && this.points.length < animationLimit) { // #8099
+            const blendColors = this.options.blendColors;
+
+            // Why ['']? If standard single bubble, perform animation only for
+            // one point (one graphic). String element matches in TS
+            (blendColors || ['']).forEach((_, i: number): void => {
```
Comment (TorsteinHonsi) [inline ts/Series/Bubble/BubbleSeries.ts:562]: It looks a bit weird to use a `forEach` loop just to get the index. Maybe better with a regular `for` loop?

`for (let i = 0; i < blendColors.length || 1; i++) {`

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]

File: ts/Series/Bubble/BubbleSeries.ts [hunk 14]
```diff
@@ -632,27 +632,36 @@ class BubbleSeries extends ScatterSeries {
 
     }
 
-    public getPxExtremes(): BubblePxExtremes {
+    public getPxSize(
+        length: number|string|undefined,
+        isMinSize?: boolean
+    ): number {
```
[the remark this answers is not in the document]
Comment (raf18seb) [reply ts/Series/Bubble/BubbleSeries.ts]: I needed that in temperaturemap wrap, but no longer needed in bubble file. I brought back old code

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Last missing piece is a map demo that is better suited to display this feature

**Review by KacperMadej — APPROVED** [review]

**Comment by TorsteinHonsi** [comment]
For the records, we decided to pivot this to try and hook the colors up to the `colorAxis`

**Comment by TorsteinHonsi** [comment]
Closing this as a dead end. We decided to go for `geoheatmap` instead.

**Pull request closed without merging**